File: 0112-path-sum/0112-path-sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean hasPathSum(TreeNode root, int targetSum) {
        if(root==null)return false;
        if(root.left ==null && root.right == null){
            if(targetSum == root.val)return true;
        }
        int sumSoFar=0;
        boolean ans = hasPathSumHelper(root, targetSum, sumSoFar);
        return ans;
    }
    public boolean hasPathSumHelper(TreeNode root, int target, int sumSoFar){
        if(root ==null)return false;
        sumSoFar+=root.val;
        if(root.left ==null && root.right == null){
            if(target == sumSoFar)return true;
        }
        return hasPathSumHelper(root.left, target, sumSoFar) || hasPathSumHelper(root.right, target, sumSoFar);
    }
}

File: 0112-path-sum/0112-path-sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:
        if not root:
            return False
        def check(root, target) -> bool:
            if not root:
                return False 
            if not root.left and not root.right:
                return target == root.val
            target -= root.val
            return check(root.left, target) or check(root.right, target)
        return check(root, targetSum)
        